soccer/src/soccer/planning/primitives/create_path.cpp

```diff
@@ -75,4 +75,94 @@ Trajectory rrt(const LinearMotionInstant& start, const LinearMotionInstant& goal
     return path;
 }
 
+Trajectory intermediate(const LinearMotionInstant& start, const LinearMotionInstant& goal,
+                        const MotionConstraints& motion_constraints, RJ::Time start_time,
+                        const rj_geometry::ShapeSet& static_obstacles) {
+    // if already on goal, no need to move
+    if (start.position.dist_to(goal.position) < 1e-6) {
+        return Trajectory{{RobotInstant{Pose(start.position, 0), Twist(), start_time}}};
+    }
+
+    // maybe straight line works
+    Trajectory straight_trajectory =
+        CreatePath::simple(start, goal, motion_constraints, start_time);
+
+    // If we are very close to the goal (i.e. there physically can't be a robot
+    // in our way) or the straight trajectory is feasible, we can use it.
+    if (start.position.dist_to(goal.position) < kRobotRadius ||
+        (!trajectory_hits_static(straight_trajectory, static_obstacles, start_time, nullptr))) {
+        return straight_trajectory;
+    }
+
+    // Generate list of intermediate points
+    std::vector<rj_geometry::Point> intermediates = get_intermediates(start, goal);
+
+    for (int i = 0; i < NUM_INTERMEDIATES; i++) {
+        rj_geometry::Point final_inter = intermediates[i];
+
+        // Step through the path from the robot to the final intermediate point
+        // and test each point on that path as an intermediate point
+        for (double t = STEP_SIZE; t < final_inter.dist_to(start.position); t += STEP_SIZE) {
+            rj_geometry::Point intermediate =
+                (final_inter - start.position).normalized(t) + start.position;
+            Trajectory trajectory =
+                CreatePath::simple(start, goal, motion_constraints, start_time, {intermediate});
+
+            // If the trajectory does not hit an obstacle, it is valid
+            if ((!trajectory_hits_static(trajectory, static_obstacles, start_time, nullptr))) {
+                return trajectory;
+            }
+        }
+    }
+
+    // If all else fails, return the straight-line trajectory
+    return straight_trajectory;
+}
+
+std::vector<rj_geometry::Point> get_intermediates(const LinearMotionInstant& start,
+                                                  const LinearMotionInstant& goal) {
+    std::random_device rd;
```

Comment:
Add intermediate that works to list of new random ones. Cache that so know have one that works. (only cache if know works). Cache means adding to array of possibilities so still find optimal one. This helps because might be finding worse random positions later and veering off path.